# 4.7 Deriving New Variables

1. Importing Libraies and Data

2. Creating Price_Label and Busiest_Day columns

3. Creating Busiest_Days and Slowest_Days columns

4. Creating Busiest_Period_of_Day column

5. Export dataframe to Prepared Folder file

In [1]:
# Import Libraries
import pandas as pd
import numpy as np 
import os

2. Import Data

In [2]:
path = r'C:\Users\rpa21\Instacart Basket Analysis'

In [45]:
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

Create price_label column

In [46]:
# Create a subset
df = df_ords_prods_merge[:1000000]

In [47]:
df.shape

(1000000, 14)

In [20]:
# Define a function

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return np.nan

In [24]:
# Apply the function

df['price_range'] = df.apply(price_label, axis=1)

C:\Users\rpa21\AppData\Local\Temp\ipykernel_11976\93659844.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [25]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    652638
Low-range product    338018
High range             9344
Name: count, dtype: int64

In [27]:
# Check most expensive product
df['prices'].max()

24.5

In [85]:
# Price label function conditions usign the loc() function

In [39]:
df.loc[(df['prices'] > 15), 'price_range_loc'] = 'High range product'

In [37]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid range product' 

In [38]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low range product'

In [40]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid range product     652638
Low range product     338018
High range product      9344
Name: count, dtype: int64

In [41]:
df.shape


(1000000, 15)

In [48]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [50]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5),  'price_range_loc'] = 'Mid-range product'

In [52]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

In [ ]:
#Checking the orders by day of week 
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

In [61]:
result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [ ]:
result

In [71]:
df_ords_prods_merge['Busiest_day'] = result

In [72]:
df_ords_prods_merge['Busiest_day'].value_counts(dropna = False)

Busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

Task Exercises

 3. Busiest Days and Slowest Days Column

In [73]:
#Create a column for the two busiest days and two slowest days of the week. 

result_2 = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value == 1:
    result_2.append("busiest days")
  elif value == 4 or value == 3:
    result_2.append("slowest days")
  else:
    result_2.append("regularly busy")

In [ ]:
result_2

In [76]:
df_ords_prods_merge['busiest days'] = result_2

In [77]:
df_ords_prods_merge['busiest days'].value_counts(dropna = False)

busiest days
regularly busy    12916111
busiest days      11864412
slowest days       7624336
Name: count, dtype: int64

3. Check the values of the column for accuracy.

In [ ]:
df_ords_prods_merge.head()

The new columns are present and the two busiest days 11,864,412 = the two busiest days of the week, '0' Saturday = 6,204,182 and '1' Sunday = 5,660,230.

4. Create a new column with labels "Most orders", "Average orders", "Fewest orders" called "busiest period of day".

In [ ]:
# Check order_hour_of_day value counts

df_ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

In [80]:
# Create the labels.

result_3 = []

for value in df_ords_prods_merge["order_hour_of_day"]:
  if value in [9, 10, 11, 12, 13, 14, 15, 16]:
    result_3.append("Most_orders")
  elif value in [0, 1, 2, 3, 4, 5, 6, 23]:
    result_3.append("Fewest_orders")
  else:
    result_3.append("Average_orders")

In [ ]:
result_3

In [82]:
# Create a new column "busiest_period_of_day".

df_ords_prods_merge["busiest_period_of_day"] = result_3

In [83]:
# Check the column.
df_ords_prods_merge["busiest_period_of_day"].value_counts(dropna = False)

busiest_period_of_day
Most_orders       21118071
Average_orders     9997651
Fewest_orders      1289137
Name: count, dtype: int64

In [ ]:
# Check the new column.
df_ords_prods_merge.head()

In [89]:
# Export as pkl file
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_new.pkl'))